File: src/rutils.R
library(edgeR)
library(MCMCpack) #for sampling from Dirichlet
library(pracma) #for convolution
library(Rsolnp) # for optimization with lagrange mult (eq constraint)

source('./mark_estdistrpar.R')


estimateTF <- function(data0, muTs, sizeTs) {

  funknown <- 1:dim(data0)[1]

  fests <- estfPGF(dat = t(data0), is = funknown, muTs = muTs, dispTs = sizeTs,
                   ngrid=25, sum2one=T, upper=1, distr= "nb")


  # Retrieve fractions
  fsremain <- unlist(lapply(fests, function(el) return(1/(1+exp(-el$par)))))

  estimatedTF <- 1-fsremain
  return(estimatedTF)
}


getDispersions <- function(dat, priorDF){
    d <- DGEList(counts = dat) #using edgeR
    d <- estimateDisp(d)

    return(list(common=d$common.dispersion, tagwise=d$tagwise.dispersion, trended=d$trended.dispersion))

}



estimateMeanVariance <- function(dataRef, label) {
  # estimate mean and variance of HBDs and tumors
  fssingle <- matrix(0, nrow=dim(dataRef)[1], 2)
  fssingle[label==0,1] = 1
  fssingle[label==1,2] = 1

  fssingle <- t(fssingle)
  dataRef <- t(dataRef)

  nT <- nrow(fssingle)
  nsingle <- ncol(fssingle)

  datsingle <- round(dataRef,0) #data should be integers

  allmus <- alldisps <- c()
  for(t in 1:nT){
    #t<-1
    whin <- which(fssingle[t,]==1)
    dat <- datsingle[,whin]
    mus <- apply(dat,1,mean)
    allmus <- rbind(allmus,mus)
    d <- DGEList(counts = dat) #using edgeR
    d.CR <- estimateCommonDisp(d)
    #if(nrow(dat) >= 100) d.CR <- estimateTrendedDisp(d.CR)
    d.CR <- estimateTagwiseDisp(d.CR)
    disps <- d.CR$tagwise.dispersion
    #var2 <- mus + mus^2*disps
    alldisps <- rbind(alldisps,disps)
  }
  muTs <- t(log(allmus+0.5))
  sizeTs <- t(alldisps)

  return(list(mus=muTs,disps=sizeTs))
}


dmrCallEdgeR <- function(counts, librarySize, label, maxGenes, minFC) {
  # counts in sklearn format, samples x features


  # put data into edgeR, normalize and do differential expression
  y <- DGEList(counts=t(counts), lib.size=librarySize, group=label)
  y <- calcNormFactors(y)

  # print(sort(y[[2]]$norm.factors))

  design <- model.matrix(~label)

  y <- estimateDisp(y,design)

  fit <- glmQLFit(y,design)
  # plotQLDisp(fit)


  dmr <- glmQLFTest(fit, coef=2)

  dmrInfo <- as.data.frame(topTags(dmr, n=dim(counts)[2], adjust.method = 'bonferroni', p.value = 0.05)$table)
  dmrInfo <- dmrInfo[abs(dmrInfo$logFC) > minFC,]

  if (dim(dmrInfo)[1] > maxGenes) {
    dmrInfo <- dmrInfo[1:maxGenes,]
  }

  return(dmrInfo)
}


dmrCallEdgeRstandard <- function(counts, librarySize, label, removeLowCounts, covariates) {
  # counts in sklearn format, samples x features
  # library size = Used reads
  # label 0 1, only binary
  # removeLowCounts: bool whether to call filterByExpr
  # covariates: things to include in the samples argument of DGEList

  print(paste0('Starting DMR calling, ', dim(counts)[1], ' samples, ', dim(counts)[2], ' features'))



  # put data into edgeR, normalize and do differential expression
  if (!is.null(dim(covariates))){
    useCovariates = TRUE
    print(paste('Including', dim(covariates)[2], 'covariates'))
    y <- DGEList(counts=t(counts), lib.size=librarySize, group=label, samples=covariates)

  } else {
    useCovariates = FALSE
    print('No or invalid covariates specified. Using no covariates')
    y <- DGEList(counts=t(counts), lib.size=librarySize, group=label)

  }


  if (removeLowCounts) {
    keep <- filterByExpr(y)
    y <- y[keep, , keep.lib.sizes=TRUE]
    print(paste('Removing regions with low counts.', sum(keep), 'regions remaining.'))
  }

  y <- calcNormFactors(y)

  # print(sort(y[[2]]$norm.factors))
  if (useCovariates) {
    design <- model.matrix(as.formula(paste("~ y$samples$group + ", paste0('y$samples$',colnames(y$samples[,-c(1,2,3),drop=FALSE]), collapse=" + "))))

  } else {
    design <- model.matrix(~label)
  }

  #print(dim(design))
  print(colnames(design))
  y <- estimateDisp(y,design)

  fit <- glmQLFit(y,design)
  # plotQLDisp(fit)


  dmr <- glmQLFTest(fit, coef=2)


  print(paste(dim(topTags(dmr, n=dim(counts)[2], adjust.method = "none", p.value=0.05))[1], "significant before correction"))
  dmrInfo <- as.data.frame(topTags(dmr, n=dim(counts)[2], adjust.method = 'fdr', p.value = 1.0)$table)
  print(paste(dim(dmrInfo)[1], "significant after correction"))
  print("\n\n")


  return(dmrInfo)
}

File: src/rutils.py
import numpy as np
import pandas as pd
import rpy2.robjects as robjects
from rpy2.robjects import numpy2ri
from rpy2.robjects import default_converter
from rpy2.robjects import pandas2ri
import pickle
import os
import sys

def getRegionLength(region: str) -> int:
    rr = region.split(':')[1].split('-')
    return int(rr[1]) - int(rr[0])

def counts2tpm(data: pd.DataFrame) -> pd.DataFrame:
    # dataframe features x samples
    # with index being chr:start-end
    # return a dataframe of the same size with tpm values

    regionSize = pd.Series(data.index).apply(getRegionLength)
    regionSize = pd.Series(np.array(regionSize), index=data.index)

    countsPerBase = data.T / regionSize
    total = countsPerBase.sum(axis=1)
    tpm = 1e6 * countsPerBase.T / total

    return tpm


def isAutosomal(region):
    autosomals = set([('chr%d' % i) for i in range(1,23)])

    return region.split(':')[0] in autosomals

def r_dmr_edgeR_standard(counts, librarysize, labels, removeLowCounts, covariates=None):
    """
    use edgeR to call DMRs, then perform FDR correction
    counts: a pd.DataFrame with shape (samples, genes)
    librarysize and labels should be np.ndarray's of shape (samples,)
    labels is assumed to have two different values (ie this finds DMRs between exactly 2 groups)
    removeLowCounts: bool whether to call filterByExpr
    covariates: things to include in the samples argument of DGEList

    # RETURNS: data frame with significant regions as rows """

    if covariates is None:
        covariates = 0

    with np_cv_rules.context():
        res = robjects.globalenv['dmrCallEdgeRstandard'](counts, librarysize, labels, removeLowCounts, covariates)


    return res



def r_dmr_edgeR_tissue(counts, librarysize, labels, maxMarkers=500, minFC=2.5):
    # use edgeR to call DMRs, then perform Bonferroni correction
    # at most maxMarkers genes are selected if significant at FWER=0.05 AND absolute foldchange > minFC
    # counts: a pd.DataFrame with shape (samples, genes)
    # librarysize and labels should be np.ndarray's of shape (samples,)
    # labels is assumed to have two different values (ie this finds DMRs between exactly 2 groups)

    # RETURNS: TODO!

    with np_cv_rules.context():
        res = robjects.globalenv['dmrCallEdgeR'](counts, librarysize, labels, maxMarkers, minFC)


    return res



def r_estimateMeanVariance(data, labels):
    with np_cv_rules.context():
        res = robjects.globalenv['estimateMeanVariance'](data, labels)


    return res


def r_estimateTF(data, mus, disps):
    with np_cv_rules.context():
        res = robjects.globalenv['estimateTF'](data, mus, disps)

    return res


def r_parseRemain(data, knownfracs, mus, disps, columnInd):

    fr = np.vstack((1-knownfracs, knownfracs))

    with np_cv_rules.context():
        res = robjects.globalenv['supervisedDeconvolution'](data, fr, mus, disps, columnInd)

    return res


def r_dispersions(data, df=10):
    with np_cv_rules.context():
        res = robjects.globalenv['getDispersions'](data, df)

    return res

np_cv_rules = default_converter + numpy2ri.converter + pandas2ri.converter

r_source = robjects.r['source']
r_source('rutils.R')







if __name__ == '__main__':
    raise NotImplementedError
